lib/cn/kvs_mblk_desc.c

```diff
@@ -71,38 +75,36 @@ mblk_munmap(struct mpool *mp, struct kvs_mblk_desc *md)
 }
 
 merr_t
-mblk_madvise(const struct kvs_mblk_desc *md, size_t off, size_t len, int advice)
+mblk_madvise_pages(const struct kvs_mblk_desc *md, size_t pg, size_t pg_cnt, int advice)
 {
-    int rc;
-    const size_t wlen = md->wlen_pages * PAGE_SIZE;
+    const size_t wlen_pages = md->wlen_pages;
+    size_t ra_pages;
+    size_t chunk = 0;
 
-    assert(md->map_base);
+    if (pg > wlen_pages)
+        return merr(EINVAL);
 
-    if (len == 0)
+    if (pg + pg_cnt > wlen_pages)
+        pg_cnt = wlen_pages - pg;
+
+    if (pg_cnt == 0)
         return 0;
 
-    if (off >= wlen)
-        return merr(EINVAL);
+    ra_pages = (advice == MADV_WILLNEED) ? md->ra_pages : pg_cnt;
 
-    if (off + len > wlen)
-        len = wlen - off;
+    for (size_t pg_end = pg + pg_cnt; pg < pg_end; pg += chunk) {
+        int rc;
 
-    /* Cast away the const of map_base for madvise().
-     */
-    rc = madvise((void *)md->map_base + off, len, advice);
-    if (rc)
-        return merr(errno);
+        chunk = min_t(size_t, pg_end - pg, ra_pages);
 
-    return 0;
-}
-
-merr_t
-mblk_madvise_pages(const struct kvs_mblk_desc *md, size_t pg, size_t pg_cnt, int advice)
-{
-    if (pg_cnt == 0)
-        return 0;
+        /* Cast away the const of map_base for madvise().
+         */
+        rc = madvise((void *)md->map_base + (pg * PAGE_SIZE), chunk * PAGE_SIZE, advice);
+        if (rc)
+            return merr(errno);
+    }
```

Comment:
Much thanks to @nabeelmmd for remember that we needed to do this and for the patch!  I reworked his patch to get rid of mblk_madvise() as it just added additional overhead and quietly failed miserably in the WILLNEED case.